File: 3.cameraandmatrices/webgl/source/js/engine/rendering/CCModelText.js
/*-----------------------------------------------------------
 * 2c - Cross Platform 3D Application Framework
 *-----------------------------------------------------------
 * Copyright © 2011 - 2011 France Telecom
 * This software is distributed under the Apache 2.0 license.
 * http://www.apache.org/licenses/LICENSE-2.0.html
 *-----------------------------------------------------------
 * File Name   : CCModelText.js
 * Description : Represents a 3d text primitive.
 *
 * Created     : 23/10/11
 * Author(s)   : Ashraf Samy Hegab
 *-----------------------------------------------------------
 */

function CCModelText(inParent)
{
	this.construct();
	
    this.primitive = false;
	this.setParent( inParent );
	this.shader = "alphacolour";
	
	this.colourInterpolator = new CCInterpolatorColour();
	this.colourInterpolator.setDuration( 0.5 );
}
CopyPrototype( CCModelText, CCModelBase, "CCModelBase" );


CCModelText.prototype.setParent = function(inParent)
{
	this.parent = inParent;
    if( inParent )
	{
		if( inParent.model == false )
		{
			inParent.model = this;
		}
		else
		{
			inParent.model.addModel( this );
		}
	}
}


CCModelText.prototype.getText = function()
{
	return this.primitive.text;
}


CCModelText.prototype.getWidth = function()
{
	return this.primitive.getWidth();
}


CCModelText.prototype.setText = function(text, height, font)
{
	if( this.primitive == false )
	{
		this.translate( 0.0, 0.0, CC_SMALLFLOAT );
		this.setColour( gColour.setGrey( 0.0, 1.0 ) );
		this.primitive = new CCPrimitiveText( text );
		this.addPrimitive( this.primitive );
	}
	
	if( height )
	{
		this.setHeight( height );
	}
	
	if( font )
	{
		this.setFont( font );
	}
	else
	{
		this.setFont( "HelveticaNeueLight" );
	}
}


CCModelText.prototype.setHeight = function(height)
{
	this.primitive.setHeight( height );
}


CCModelText.prototype.setCentered = function(centered)
{
	if( centered )
	{
		this.setPositionX( 0.0 );
	}
	else if( this.parent )
	{
		this.setPositionX( -this.parent.collisionBounds[0] );
	}
	this.primitive.setCentered( centered );
}


CCModelText.prototype.setColour = function(inColour)
{
	this.CCModelBase_setColour( inColour );
	this.colourInterpolator.setup( this.colour );
}


CCModelText.prototype.setFont = function(font)
{
	this.primitive.setFont( font );
}



// CCPrimitiveText
function CCPrimitiveText(inText, inHeight)
{
	this.construct();
	
	this.setText( inText );
	this.setHeight( inHeight );
	this.setCentered( true );
}
CopyPrototype( CCPrimitiveText, CCPrimitiveBase, "CCPrimitiveBase" );


CCPrimitiveText.prototype.renderVertices = function(renderer)
{
	var text = this.text;
	this.fontPage.renderText3D( text, text.length, 0.0, 0.0, 0.0, this.height, this.centered );
}


CCPrimitiveText.prototype.setText = function(inText)
{
	this.text = inText;
}


CCPrimitiveText.prototype.setFont = function(inFont)
{
	var fontPages = gEngine.textureManager.fontPages;
	var length = fontPages.length;
	for( var i=0; i<length; ++i )
	{
		var page = fontPages[i];
		if( inFont == page.name )
		{
			this.fontPage = page;
			return;
		}
	}
}


CCPrimitiveText.prototype.setHeight = function(inHeight)
{
	this.height = inHeight;
}


CCPrimitiveText.prototype.setCentered = function(inCentered)
{
	this.centered = inCentered;
}


CCPrimitiveText.prototype.getWidth = function()
{
	var text = this.text;
	return this.fontPage.getWidth( text, text.length, this.height );
}

File: 3.cameraandmatrices/mobile/Engine/Source/Rendering/CCModelText.cpp
/*-----------------------------------------------------------
 * 2c - Cross Platform 3D Application Framework
 *-----------------------------------------------------------
 * Copyright © 2010 – 2011 France Telecom
 * This software is distributed under the Apache 2.0 license,
 * see the "license.txt" file for more details.
 *-----------------------------------------------------------
 * File Name   : CCModelText.cpp
 *-----------------------------------------------------------
 */

#include "CCDefines.h"
#include "CCObjects.h"
#include "CCTextureManager.h"
#include "CCTextureFontPage.h"


CCModelText::CCModelText(CCSceneCollideable *inParent)
{
    primitive = NULL;
    parent = NULL;
    setParent( inParent );
    
    shader = "alphacolour";
    
    colourInterpolator.setDuration( 0.5f );
}


void CCModelText::destruct()
{
    super::destruct();
}


void CCModelText::setParent(CCSceneCollideable *inParent)
{
    ASSERT( parent == NULL );
    parent = inParent;
    if( inParent != NULL )
    {
        if( parent->model == NULL )
        {
            parent->model = this;
        }
        else
        {
            parent->model->addModel( this );
        }
    }
}


void CCModelText::setText(const char *text, const float height, const char *font)
{
    if( primitive == NULL )
    {
        translate( 0.0f, 0.0f, CC_SMALLFLOAT );
        setColour( CCColour( 0.0f ) );
        primitive = new CCPrimitiveText( text );
        addPrimitive( primitive );
    }
    else
    {
        primitive->setText( text );
    }
    
    if( height != -1.0f )
    {
        setHeight( height );
    }
    
    if( font != NULL )
    {
        setFont( font );
    }
    else
    {
        setFont( "HelveticaNeueLight" );
    }
}


void CCModelText::setHeight(const float height)
{
    primitive->setHeight( height );
}


void CCModelText::setCentered(const bool centered)
{
    if( centered )
    {
        setPositionX( 0.0f );
    }
    else if( parent != NULL )
    {
        setPositionX( -parent->collisionBounds.x );
    }
    primitive->setCentered( centered );
}


void CCModelText::setColour(const CCColour &inColour)
{
    super::setColour( inColour );
    colourInterpolator.setup( colour, *colour );
}


void CCModelText::setFont(const char *font)
{
    primitive->setFont( font );
}



// CCPrimitiveText
CCPrimitiveText::CCPrimitiveText(const char *inText, const float inHeight)
{	
    setText( inText );
    setHeight( inHeight );
    setCentered( true );
}


void CCPrimitiveText::destruct()
{	
	super::destruct();
}


// CCPrimitiveBase
void CCPrimitiveText::renderVertices(const bool textured)
{	
    if( text.buffer != NULL )
    {
        fontPage->renderText3D( text.buffer, text.length, 
                                0.0f, 0.0f, 0.0f,
                                height,
                                centered );
    }
}


void CCPrimitiveText::setText(const char *inText)
{
    text = inText;
}


void CCPrimitiveText::setFont(const char *font)
{
    for( int i=0; i<gEngine->textureManager->fontPages.length; ++i )
    {
        CCTextureFontPage *page = gEngine->textureManager->fontPages.list[i];
        const char *name = page->getName();
        if( CCText::Equals( font, name ) )
        {
            fontPage = page;
            return;
        }
    }
    ASSERT( false );
}


const float CCPrimitiveText::getWidth()
{
    return fontPage->getWidth( text.buffer, text.length, height );
}

